### DISCUSSION TITLE
Can rehypeRemark passthrough all non standard html tags?

### DISCUSSION THREAD

**Discussion by giorgio-zamparelli** [rehypejs/rehype#125] [Q&A] [answered]
I'm using rehype-remark to convert some html into markdown but I need to keep custom html tags who I don't know in advance as they are.

<note> could be any non standard html tag and it's not known in advance.

INPUT HTML
```<h1>title</h1>\n<note>some note here</note>```

WANTED OUTPUT MARKDOWN
```# title\n<note>some note here</note>

rehype-remark has the option handlers to specify how to serialize a specific tag.
The problem is do NOT know the name of the custom-tag since it is defined by the user and I wish to passthrough all non-standard html tags.

```unified()
.use(rehypeParse, { fragment: true }) // Parse HTML to a syntax tree
.use(rehypeRemark, {
  handlers: {
    note(h, node) {
      return h(node, 'html', toHtml(node))
    },
  },
}) // Turn HTML syntax tree to markdown syntax tree
.use(remarkStringify) // Serialize HTML syntax tree
.process(html)```

**Comment by ChristianMurphy** [comment, marked as answer]
Welcome @giorgio-zamparelli! 👋 

Some expectation setting before diving into things.
Your example, does not contain custom HTML tags.
Custom elements _must_ have a dash in their name (see https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements and https://html.spec.whatwg.org/#valid-custom-element-name)
`<note>` is not a built in tag or a custom element, it is invalid HTML.

Most likely you're trying to leverage another language like: JSX, Angular templates, Glimmer templates, etc.
If you need your content to be valid in a specific templating language, use a parser designed for that language.
HTML parsers will go into a recovery mode to try to handle this, which will result in a valid HTML page, but will likely be invalid for a different templating language.

Back to your direct question:

> I need to keep custom html tags who I don't know in advance as they are.

Use a `Proxy` to create `handlers` to catch all elements (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy), and combine it with https://www.npmjs.com/package/html-tag-names to skip handling built-in elements, and handle what remains (custom elements) with a handler similar to what you show for note above.
